daemon/api_snaps.go

```diff
@@ -684,18 +684,19 @@ func snapEnforceValidationSets(inst *snapInstruction, st *state.State) (*snapIns
 		return nil, err
 	}
 
-	var validationErr *snapasserts.ValidationSetsValidationError
+	var tss []*state.TaskSet
+	var affected []string
 	err = assertstateTryEnforceValidationSets(st, inst.ValidationSets, inst.userID, snaps, ignoreValidationSnaps)
 	if err != nil {
-		var ok bool
-		validationErr, ok = err.(*snapasserts.ValidationSetsValidationError)
+		vErr, ok := err.(*snapasserts.ValidationSetsValidationError)
 		if !ok {
 			return nil, err
 		}
-	}
-	tss, affected, err := snapstateEnforceSnaps(context.TODO(), st, inst.ValidationSets, validationErr, inst.userID)
-	if err != nil {
-		return nil, err
+
+		tss, affected, err = meetSnapConstraintsForEnforce(inst, st, vErr)
+		if err != nil {
+			return nil, err
+		}
 	}
 
 	return &snapInstructionResult{
```

Comment:
I didn't notice before but the summary here doesn't look like a Change summary, it should be something like

"Enforce validation sets %s for snaps %s"

you might need to tweak if affected is empty?